raft/raft.go

```diff
@@ -1211,6 +1211,16 @@ func (r *raft) handleSnapshot(m pb.Message) {
 // restore recovers the state machine from a snapshot. It restores the log and the
 // configuration of state machine.
 func (r *raft) restore(s pb.Snapshot) bool {
+	// The normal peer can't become learner.
+	if !r.isLearner {
```

Comment:
move this checking below the raft log checking. we should check normal raft stuff first to make code easier to read.